source/plugin/event/causes.rst

```diff
@@ -69,55 +99,164 @@ object matching the provided type.
+This example makes use of :javadoc:`EventContext#get(EventContextKey)` which can be used to retrieve the expected object
+associated with a name if it is present within the context. Additionally :javadoc:`EventContext#asMap()` provides
+a ``Map<EventContextKey<?>, Object>`` which can be used to find all present ``EventContextKey``\s and their associated
+objects.
 
 .. note::
 
-    Some common identifying names for ``NamedCause``\ s are present as static fields in the
-    ``NamedCause`` class. Identifiers which are specific to certain events can often be found
-    as static fields on the event class, for example :javadoc:`DamageEntityEvent#SOURCE`.
+    Some common identifying names for ``EventContextKey``\s are present as static fields in the
+    ``EventContextKeys`` class.
 
 Creating custom Causes
 ~~~~~~~~~~~~~~~~~~~~~~
 
-Creating a cause to use when firing an event is extremely easy. The hardest part is deciding
-what information to include in the cause. If you're firing an event from your plugin which is
-usually triggered through other means perhaps you want to include your plugin container so
-other plugins know that the event comes from your plugin. Or if you are firing the event on
-behalf of a player due to some action it's usually a good idea to include that player in
-the cause.
+Creating a cause is easy, but depends on whether you are creating your cause on the main server
+thread or async.
 
 .. note::
 
     Cause objects are immutable therefore cannot be modified once created.
 
-Using :javadoc:`Cause#of(NamedCause)`, you can construct a cause from a series of objects. The objects will be added to
-the cause chain in the order that they are passed to the method, so the first object parameter will become the root
-cause. Remember that a ``Cause`` may not be empty, so at least one non-null parameter is always required.
+Using the CauseStackManager
+===========================
+
+.. warning::
+
+    The ``CauseStackManager`` only works on the main server thread. If you call it from a
+    different thread, an ``IllegalStateException`` will be thrown. Ensure you are on the main
+    server thread **before** calling methods on the ``CauseStackManager``.
+
+If you are creating your event on the main thread, then use the :javadoc:`CauseStackManager`, which can
+be found at :javadoc:`Sponge#getCauseStackManager()`. The ``CauseStackManager`` tracks the potential
+causes of events as the game runs, allowing for easy retrieval of the current ``Cause`` without effort.
+To see the current cause, call :javadoc:`CauseStackManager#getCurrentCause()`. You may notice that your
+plugin's :javadoc:`PluginContainer` is already in the returned ``Cause``, as plugins are one of the
+objects tracked by the manager. Using the ``CauseStackManager`` for creating causes removes the
+need for boilerplate-like code where you supply objects like your plugin container, you can concentrate
```

Comment:
This sentence is incomplete, it needs a "so". i.e. "... container, *so* you can concentrate ..."